KFSPTS-9311 prevent duplicate entries from GEC docs into CAM

jhulslander opened the pull request.

db483 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -408,8 +414,20 @@ protected boolean doesGeneralLedgerEntryMatchAssetAccountingDetails( GeneralLedg
         //check if GL Debit matches acct line type target and Credit matches Source
-		if ((StringUtils.equals(entry.getTransactionDebitCreditCode(), KFSConstants.GL_DEBIT_CODE) && StringUtils.equals(accountingDetails.getFinancialDocumentLineTypeCode(), KFSConstants.TARGET_ACCT_LINE_TYPE_CODE))
-				|| (StringUtils.equals(entry.getTransactionDebitCreditCode(), KFSConstants.GL_CREDIT_CODE) && StringUtils.equals(accountingDetails.getFinancialDocumentLineTypeCode(), KFSConstants.SOURCE_ACCT_LINE_TYPE_CODE))) {
+		boolean isGlEntryDebit = StringUtils.equals(entry.getTransactionDebitCreditCode(), KFSConstants.GL_DEBIT_CODE);
```
there seems to be a formatting issue on this line

db483 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -408,8 +414,20 @@ protected boolean doesGeneralLedgerEntryMatchAssetAccountingDetails( GeneralLedg
+        boolean isAccountingDetailTarget = StringUtils.equals(accountingDetails.getFinancialDocumentLineTypeCode(), KFSConstants.TARGET_ACCT_LINE_TYPE_CODE);
+        boolean isAccountingDetailSource = StringUtils.equals(accountingDetails.getFinancialDocumentLineTypeCode(), KFSConstants.SOURCE_ACCT_LINE_TYPE_CODE);
+        boolean isGeneralErrorCorrection = StringUtils.equalsIgnoreCase(entry.getFinancialDocumentTypeCode(), KFSConstants.FinancialDocumentTypeCodes.GENERAL_ERROR_CORRECTION);
+        if (LOG.isDebugEnabled()) {
```
I think all there logging statements could be removed now that development is done. They clutter the code more than they help at this point.

jhulslander replied:
How about moving these logs to a separate function, that will keep this function cleaner, I think this log statement will be helpful

db483 replied:
that makes sense, thank you

db483 commented on src/main/java/org/kuali/kfs/module/cam/web/struts/CapitalAssetInformationAction.java:
```diff
@@ -103,6 +110,64 @@ protected void prepareRecordsForDisplay(CapitalAssetInformationForm capitalAsset
+            Collection<GeneralLedgerEntry> glEntries) {
+        KualiDecimal entriesTotal = findTotalGLEntries(glEntries);
+        KualiDecimal informationTotal = findTotalAmountForAssetInformation(capitalAssetInformations);
+        if (LOG.isDebugEnabled()) {
```
I think the code would be cleaner and more clear if you remove the debug logging.

jhulslander replied:
How about moving these logs to a separate function, that will keep this function cleaner, I think this log statement will be helpful

db483 commented on src/main/java/org/kuali/kfs/module/cam/web/struts/CapitalAssetInformationAction.java:
```diff
@@ -103,6 +110,64 @@ protected void prepareRecordsForDisplay(CapitalAssetInformationForm capitalAsset
+            isGeneralErrorCorrection = StringUtils.equalsIgnoreCase(firstEntry.getFinancialDocumentTypeCode(), 
+                    KFSConstants.FinancialDocumentTypeCodes.GENERAL_ERROR_CORRECTION);
+        }
+        if (LOG.isDebugEnabled()) {
```
I think the code would be cleaner and more clear if you remove the debug logging.

jhulslander replied:
sure

db483 commented on src/main/java/org/kuali/kfs/module/cam/web/struts/CapitalAssetInformationAction.java:
```diff
@@ -103,6 +110,64 @@ protected void prepareRecordsForDisplay(CapitalAssetInformationForm capitalAsset
+        KualiDecimal amount = KualiDecimal.ZERO;
+        if (CollectionUtils.isNotEmpty(informationList)) {
+            for (CapitalAssetInformation info : informationList) {
+                amount = amount.abs().add(info.getCapitalAssetLineAmount().abs());
```
I am not sure I understand why you add the absolute values here. Are you sure this is going to work without taking into account debits and credits and positives and negatives?

jhulslander replied:
let me try switching this around.

db483 commented on src/main/java/org/kuali/kfs/module/cam/web/struts/CapitalAssetInformationAction.java:
```diff
@@ -103,6 +110,64 @@ protected void prepareRecordsForDisplay(CapitalAssetInformationForm capitalAsset
+    private KualiDecimal findTotalAmountForAssetInformation(List<CapitalAssetInformation> informationList) {
+        KualiDecimal amount = KualiDecimal.ZERO;
+        if (CollectionUtils.isNotEmpty(informationList)) {
+            for (CapitalAssetInformation info : informationList) {
```
I wonder if you could implement this using streams and reduce.

jhulslander replied:
good idea

db483 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -129,13 +130,18 @@ protected void markCapitalAssetProcessed(GeneralLedgerEntry primary, Integer cap
         if (ObjectUtils.isNotNull(capitalAssetInformation)) {
             List<CapitalAssetAccountsGroupDetails> groupAccountingLines = capitalAssetInformation.getCapitalAssetAccountsGroupDetails();
             Collection<GeneralLedgerEntry> documentGlEntries = findAllGeneralLedgerEntry(entry.getDocumentNumber());
+            if (LOG.isDebugEnabled()) {
```
I think the code would be cleaner and more clear if you remove the debug logging.

jhulslander replied:
sure, this can be removed

db483 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -129,13 +130,18 @@ protected void markCapitalAssetProcessed(GeneralLedgerEntry primary, Integer cap
             for (CapitalAssetAccountsGroupDetails accountingLine : groupAccountingLines) {
                 //find the matching GL entry for this accounting line.
                 Collection<GeneralLedgerEntry> glEntries = findMatchingGeneralLedgerEntries(documentGlEntries, accountingLine);
+                if (LOG.isDebugEnabled()) {
```
I think the code would be cleaner and more clear if you remove the debug logging.

jhulslander replied:
sure, this can be removed

jhulslander commented:
Chad, Daniela asked me to get another reviewer of this PR so we can move forward while she is out.  I will assign the Jira to you.

cah292 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -273,24 +273,44 @@ public CapitalAssetInformation findCapitalAssetInformation(String documentNumber
+                boolean isGroupLineTarget = StringUtils.equals(groupAccountLine.getFinancialDocumentLineTypeCode(), KFSConstants.TARGET_ACCT_LINE_TYPE_CODE);
+                boolean isGLEntryCredit = StringUtils.equals(entry.getTransactionDebitCreditCode(), KFSConstants.GL_CREDIT_CODE);
+                boolean isGLEntryDebit = StringUtils.equals(entry.getTransactionDebitCreditCode(), KFSConstants.GL_DEBIT_CODE);
+                boolean isErrorCorrection = groupAccountLine.getAmount().isNegative();
```
Just to confirm, is this a sufficient check for seeing if the accounting line is an error correction? I'm not familiar enough with these documents to know whether non-error-correction lines are always positive (even if they're credits), so I just wanted to make sure.

cah292 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -408,16 +428,34 @@ protected boolean doesGeneralLedgerEntryMatchAssetAccountingDetails( GeneralLedg
+        boolean isGlEntryCredit = StringUtils.equals(entry.getTransactionDebitCreditCode(), KFSConstants.GL_CREDIT_CODE);
+        boolean isAccountingDetailTarget = StringUtils.equals(accountingDetails.getFinancialDocumentLineTypeCode(), KFSConstants.TARGET_ACCT_LINE_TYPE_CODE);
+        boolean isAccountingDetailSource = StringUtils.equals(accountingDetails.getFinancialDocumentLineTypeCode(), KFSConstants.SOURCE_ACCT_LINE_TYPE_CODE);
+        boolean isErrorCorrection = accountingDetails.getAmount().isNegative();
```
Another spot to verify if a negative amount is sufficient for identifying the line/details as an error correction.

jhulslander replied:
I'll add more details

cah292 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -273,24 +273,44 @@ public CapitalAssetInformation findCapitalAssetInformation(String documentNumber
     }
+    
+    private void logAddToCapitalAssetsMatchingLineTypeDetails(boolean isGroupLineSource, boolean isGroupLineTarget, boolean isGLEntryCredit, 
+            boolean isGLEntryDebit, boolean isErrorCorrection) {
```
This logging method doesn't seem to be adding quite enough value, since it prints the flags from the various checks but does not correlate them to the line that is being checked. I would suggest either adding some information about the line being examined, or just remove this method altogether.

jhulslander replied:
I'll see if I can add more details

cah292 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -408,16 +428,34 @@ protected boolean doesGeneralLedgerEntryMatchAssetAccountingDetails( GeneralLedg
     }
 
+    private void logDoesGeneralLedgerEntryMatchAssetAccountingDetails(boolean isGlEntryDebit, boolean isGlEntryCredit,
+            boolean isAccountingDetailTarget, boolean isAccountingDetailSource, boolean isErrorCorrection) {
```
Another method where it would be helpful to either add more info about the line/details being examined or to just remove the method entirely.

cah292 commented on src/main/java/org/kuali/kfs/module/cam/document/service/impl/GlLineServiceImpl.java:
```diff
@@ -273,24 +273,44 @@ public CapitalAssetInformation findCapitalAssetInformation(String documentNumber
+                logAddToCapitalAssetsMatchingLineTypeDetails(isGroupLineSource, isGroupLineTarget, isGLEntryCredit, isGLEntryDebit, isErrorCorrection);
+                if ((isGroupLineSource && isGLEntryCredit) || (isGroupLineTarget && isGLEntryDebit) || 
+                        (isErrorCorrection && isGroupLineSource && isGLEntryDebit) ||
+                        (isErrorCorrection && isGroupLineTarget && isGLEntryCredit)){
```
Missing space between the parentheses and the curly brace.

jhulslander replied:
ok

jhulslander commented:
I fixed the space and added some more logging details based on the code review.

Karen P. and I spoke about the logic to determine if we are in an error correction, and she agreed that negative amount check should be sufficient.  There are save errors on documents that prevent negative amounts when not in an error correction on the DI and GEC docs, which have been our test case documents for this bug.

this is ready for review again.
